aws-android-sdk-iot-test/src/androidTest/java/com/amazonaws/mobileconnectors/iot/IotClient.java

```diff
@@ -0,0 +1,279 @@
+import com.amazonaws.services.iot.model.DescribeEndpointRequest;
+import com.amazonaws.services.iot.model.DescribeEndpointResult;
+import com.amazonaws.services.iot.model.DetachPolicyRequest;
+import com.amazonaws.services.iot.model.ListCertificatesRequest;
+import com.amazonaws.services.iot.model.ListCertificatesResult;
+import com.amazonaws.services.iot.model.ListTargetsForPolicyRequest;
+import com.amazonaws.services.iot.model.ListTargetsForPolicyResult;
+import com.amazonaws.services.iot.model.ResourceNotFoundException;
+import com.amazonaws.services.iot.model.UpdateCertificateRequest;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Wrapper for the {@link AWSIot} client, to reduce boilerplate in test.
+ * This class should only import classes beginning com.amazonaws.services.iot,
+ * and the methods here should stick to the script wrt IoT service APIs.
+ */
+@SuppressWarnings("SameParameterValue")
+final class IotClient {
+    private static final String TAG = IotClient.class.getSimpleName();
+
+    private final AWSIot iot;
+    private final Map<String, String> endpointsCache;
+
+    IotClient(AWSCredentialsProvider credentialsProvider) {
+        this.iot = new AWSIotClient(credentialsProvider);
+        this.endpointsCache = new HashMap<>();
+    }
+
+    String getEndpointAddress() {
+        return getEndpointAddress(null);
+    }
+
+    String getEndpointAddress(@Nullable String endpointType) {
+        String cachedValue = endpointsCache.get(endpointType);
+        if (cachedValue != null) {
+            Log.d(TAG, "Retrieved cached endpoint = " + cachedValue + " for type = " + endpointType);
+            return cachedValue;
+        }
+        Log.d(TAG, "No cached endpoint of type " + endpointType + ", retrieving now.");
+
+        DescribeEndpointRequest request = new DescribeEndpointRequest();
+        if (endpointType != null) {
+            request.setEndpointType(endpointType);
+        }
+        DescribeEndpointResult result = iot.describeEndpoint(request);
+        String endpointAddress = result.getEndpointAddress();
+        endpointsCache.put(endpointType, endpointAddress);
+        return endpointAddress;
+    }
+
+    String getEndpointPrefix() {
+        return getEndpointPrefix(null);
+    }
+
+    String getEndpointPrefix(@Nullable String endpointType) {
+        return getEndpointAddress(endpointType).split("\\.")[0];
+    }
+
+    void deleteCertificate(@NonNull String certificateId) {
+        Objects.requireNonNull(certificateId);
+
+        Log.d(TAG, "Deleting certificate with ID = " + certificateId);
+
+        DeleteCertificateRequest deleteCertificateRequest = new DeleteCertificateRequest();
+        deleteCertificateRequest.setCertificateId(certificateId);
+        iot.deleteCertificate(deleteCertificateRequest);
+    }
+
+    private void updateCertificate(
```

Comment:
Is there a reason for the visibility difference here?

Reply:
I'll just make them all package-private. I don't think the current distribution of visibility makes much sense.